diff --git a/setup.py b/setup.py
index 2ea0d0b..28f2122 100644
--- a/setup.py
+++ b/setup.py
@@ -21,12 +21,21 @@ setup(
                 'cxx': ['-O3'],
                 'nvcc': [
                     '-O3',
+                    '--expt-relaxed-constexpr',
+                    '--expt-extended-lambda',
+                    '-Xcudafe',
+                    '--diag_suppress=20011',
+                    '-Xcudafe',
+                    '--diag_suppress=20012',
+                    '-Xcudafe',
+                    '--diag_suppress=20014',
                     # '-gencode=arch=compute_60,code=sm_60',
                     # '-gencode=arch=compute_61,code=sm_61',
                     '-gencode=arch=compute_70,code=sm_70',
                     '-gencode=arch=compute_75,code=sm_75',
                     '-gencode=arch=compute_80,code=sm_80',
                     '-gencode=arch=compute_86,code=sm_86',
+                    '-gencode=arch=compute_89,code=sm_89',
                 ],
             },
         ),
@@ -56,12 +65,21 @@ setup(
                 'cxx': ['-O2'],
                 'nvcc': [
                     '-O2',
+                    '--expt-relaxed-constexpr',
+                    '--expt-extended-lambda',
+                    '-Xcudafe',
+                    '--diag_suppress=20011',
+                    '-Xcudafe',
+                    '--diag_suppress=20012',
+                    '-Xcudafe',
+                    '--diag_suppress=20014',
                     # '-gencode=arch=compute_60,code=sm_60',
                     # '-gencode=arch=compute_61,code=sm_61',
                     '-gencode=arch=compute_70,code=sm_70',
                     '-gencode=arch=compute_75,code=sm_75',
                     '-gencode=arch=compute_80,code=sm_80',
                     '-gencode=arch=compute_86,code=sm_86',
+                    '-gencode=arch=compute_89,code=sm_89',
                 ],
             },
         ),
diff --git a/src/altcorr_kernel.cu b/src/altcorr_kernel.cu
index 9e0c5f7..be4f6df 100644
--- a/src/altcorr_kernel.cu
+++ b/src/altcorr_kernel.cu
@@ -304,7 +304,7 @@ std::vector<torch::Tensor> altcorr_cuda_forward(
   const dim3 threads(BLOCK_H, BLOCK_W);
 
 
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(fmap1.type(), "altcorr_forward_kernel", ([&] {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(fmap1.scalar_type(), "altcorr_forward_kernel", ([&] {
     altcorr_forward_kernel<scalar_t><<<blocks, threads>>>(
         fmap1.packed_accessor32<scalar_t,4,torch::RestrictPtrTraits>(),
         fmap2.packed_accessor32<scalar_t,4,torch::RestrictPtrTraits>(),
@@ -351,4 +351,4 @@ std::vector<torch::Tensor> altcorr_cuda_backward(
     radius);
 
   return {fmap1_grad, fmap2_grad, coords_grad};
-}
\ No newline at end of file
+}
diff --git a/src/correlation_kernels.cu b/src/correlation_kernels.cu
index 866b0b1..d84e605 100644
--- a/src/correlation_kernels.cu
+++ b/src/correlation_kernels.cu
@@ -141,7 +141,7 @@ std::vector<torch::Tensor> corr_index_cuda_forward(
   torch::Tensor corr = torch::zeros(
     {batch_size, 2*radius+1, 2*radius+1, ht, wd}, opts);
 
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(volume.type(), "sampler_forward_kernel", ([&] {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(volume.scalar_type(), "sampler_forward_kernel", ([&] {
     corr_index_forward_kernel<scalar_t><<<blocks, threads>>>(
       volume.packed_accessor32<scalar_t,5,torch::RestrictPtrTraits>(),
       coords.packed_accessor32<float,4,torch::RestrictPtrTraits>(),
@@ -172,7 +172,7 @@ std::vector<torch::Tensor> corr_index_cuda_backward(
   const dim3 threads(BLOCK, BLOCK);
 
 
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(volume.type(), "sampler_backward_kernel", ([&] {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(volume.scalar_type(), "sampler_backward_kernel", ([&] {
     corr_index_backward_kernel<scalar_t><<<blocks, threads>>>(
       coords.packed_accessor32<float,4,torch::RestrictPtrTraits>(),
       corr_grad.packed_accessor32<scalar_t,5,torch::RestrictPtrTraits>(),
@@ -181,4 +181,4 @@ std::vector<torch::Tensor> corr_index_cuda_backward(
    }));
 
   return {volume_grad};
-}
\ No newline at end of file
+}
diff --git a/thirdparty/lietorch/lietorch/extras/corr_index_kernel.cu b/thirdparty/lietorch/lietorch/extras/corr_index_kernel.cu
index 2f4652f..37e3222 100644
--- a/thirdparty/lietorch/lietorch/extras/corr_index_kernel.cu
+++ b/thirdparty/lietorch/lietorch/extras/corr_index_kernel.cu
@@ -142,7 +142,7 @@ std::vector<torch::Tensor> corr_index_cuda_forward(
   torch::Tensor corr = torch::zeros(
     {batch_size, 2*radius+1, 2*radius+1, ht, wd}, opts);
 
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(volume.type(), "sampler_forward_kernel", ([&] {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(volume.scalar_type(), "sampler_forward_kernel", ([&] {
     corr_index_forward_kernel<scalar_t><<<blocks, threads>>>(
       volume.packed_accessor32<scalar_t,5,torch::RestrictPtrTraits>(),
       coords.packed_accessor32<float,4,torch::RestrictPtrTraits>(),
@@ -173,7 +173,7 @@ std::vector<torch::Tensor> corr_index_cuda_backward(
   const dim3 threads(BLOCK, BLOCK);
 
 
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(volume.type(), "sampler_backward_kernel", ([&] {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(volume.scalar_type(), "sampler_backward_kernel", ([&] {
     corr_index_backward_kernel<scalar_t><<<blocks, threads>>>(
       coords.packed_accessor32<float,4,torch::RestrictPtrTraits>(),
       corr_grad.packed_accessor32<scalar_t,5,torch::RestrictPtrTraits>(),
diff --git a/thirdparty/lietorch/lietorch/extras/extras.cpp b/thirdparty/lietorch/lietorch/extras/extras.cpp
index 0814e32..200f462 100644
--- a/thirdparty/lietorch/lietorch/extras/extras.cpp
+++ b/thirdparty/lietorch/lietorch/extras/extras.cpp
@@ -3,7 +3,7 @@
 
 
 // C++ interface
-#define CHECK_CUDA(x) TORCH_CHECK(x.type().is_cuda(), #x " must be a CUDA tensor")
+#define CHECK_CUDA(x) TORCH_CHECK(x.is_cuda(), #x " must be a CUDA tensor")
 #define CHECK_CONTIGUOUS(x) TORCH_CHECK(x.is_contiguous(), #x " must be contiguous")
 #define CHECK_INPUT(x) CHECK_CUDA(x); CHECK_CONTIGUOUS(x)
 
diff --git a/thirdparty/lietorch/lietorch/include/dispatch.h b/thirdparty/lietorch/lietorch/include/dispatch.h
index 9c010a8..ed09410 100644
--- a/thirdparty/lietorch/lietorch/include/dispatch.h
+++ b/thirdparty/lietorch/lietorch/include/dispatch.h
@@ -36,7 +36,7 @@
   [&] {                                                                              \
     const auto& the_type = TYPE;                                                     \
     /* don't use TYPE again in case it is an expensive or side-effect op */          \
-    at::ScalarType _st = ::detail::scalar_type(the_type);                            \
+    at::ScalarType _st = the_type.scalarType();                                      \
     switch (_st) {                                                                   \
       PRIVATE_CASE_TYPE(GROUP_INDEX, at::ScalarType::Double, double, __VA_ARGS__)    \
       PRIVATE_CASE_TYPE(GROUP_INDEX, at::ScalarType::Float, float, __VA_ARGS__)      \
